[Graph Runtime] Run_individual for benchmarking individual layers

hlu1 opened the pull request:
I find this method very helpful for identifying bottleneck layers when I'm optimizing end-to-end model performance, especially when python is not available.

@tqchen, @ajtulloch, please review

grwlf commented:
Looks like an important tool. Could you please also add a test or example demonstrating how it may be used?

icemelon9 commented on src/runtime/graph/debug/graph_runtime_debug.cc:
```diff
@@ -119,6 +152,14 @@ PackedFunc GraphRuntimeDebug::GetFunction(
           this->DebugGetNodeOutput(args[0], args[1]);
         }
       });
+  } else if (name == "run_individual") {
```
Could you expose this function to python API and add a test case?

hlu1 replied:
Yes, will do

zhiics commented on src/runtime/graph/debug/graph_runtime_debug.cc:
```diff
@@ -38,6 +38,39 @@ class GraphRuntimeDebug : public GraphRuntime {
+    }
+    std::vector<double> time_per_op(op_execs().size(), 0);
+    for (int k = 0; k < iter; k++) {
+      for (size_t i = 0; i < op_execs().size(); ++i) {
```
Should we only measure operators and exclude input and weight nodes? It looks that there is no need to looping on `op_execs()[i] == nullptr`.

hlu1 replied:
You're right

hlu1 commented on src/runtime/graph/debug/graph_runtime_debug.cc:
```diff
@@ -38,6 +38,39 @@ class GraphRuntimeDebug : public GraphRuntime {
+   * \param warmup Number of warmup runs.
+   * \param iter Number of iteration runs.
+   */
+  void RunIndividual(int warmup, int iter) {
```
Yes, the latency of each op could vary a lot. Is `min_repeat_ms` for the whole model or each op? Do you find adding this parameter helpful to get consistent benchmark result? What's the typical number you use?

icemelon9 replied:
I mean `min_repeat_ms` for each op. Yes, I think it'll make the benchmark result more consistent. And it's also consistent with other time evaluator API, such as
https://github.com/dmlc/tvm/blob/master/python/tvm/module.py#L130

tqchen commented:
@hlu1 can you please fix the lint error and address the review comments? @icemelon9 please followup this PR

hlu1 commented:
@icemelon9, sorry about the delay. I changed the interface to be the same as the tvm time evaluator API. The implementation is pretty close to the time evaluator here https://github.com/dmlc/tvm/blob/master/src/runtime/rpc/rpc_session.cc#L1194. I also added python API and python tests.

tqchen commented:
@icemelon9 can you moderate this PR?

tqchen commented:
https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly@zhiics @icemelon9 please

zhiics commented on src/runtime/graph/debug/graph_runtime_debug.cc:
```diff
@@ -38,6 +38,66 @@ class GraphRuntimeDebug : public GraphRuntime {
+   * \param number The number of times to run this function for taking average.
+   * \param repeat The number of times to repeat the measurement.
+            In total, the function will be invoked (1 + number x repeat) times,
+            where the first one is warm up and will be discarded in case
```
```suggestion
            where the first one is warmed up and will be discarded in case
```

hlu1 replied:
Done

zhiics commented on src/runtime/graph/debug/graph_runtime_debug.cc:
```diff
@@ -38,6 +38,66 @@ class GraphRuntimeDebug : public GraphRuntime {
+   * \param repeat The number of times to repeat the measurement.
+            In total, the function will be invoked (1 + number x repeat) times,
+            where the first one is warm up and will be discarded in case
+            there is lazy initialization..
```
```suggestion
            there is lazy initialization.
```

hlu1 replied:
Done

tqchen commented:
@hlu1 please update per review comments, and let us get it in

tqchen commented:
Thanks, @hlu1 @icemelon9 @zhiics, this is now merged
